src/Sema.zig

```diff
@@ -22187,7 +22187,35 @@ fn zirMemset(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!void
     const dest_ptr = try sema.resolveInst(extra.lhs);
     const uncoerced_elem = try sema.resolveInst(extra.rhs);
     const dest_ptr_ty = sema.typeOf(dest_ptr);
-    try checkIndexable(sema, block, dest_src, dest_ptr_ty);
+
+    const is_memsetable = switch (dest_ptr_ty.zigTypeTag()) {
+        .Pointer => switch (dest_ptr_ty.ptrSize()) {
+            .Slice, .Many, .C => true,
+            .One => dest_ptr_ty.elemType().zigTypeTag() == .Array,
+        },
+        else => false,
+    };
+
+    if (!is_memsetable) {
+        const msg = msg: {
+            const msg = try sema.errMsg(
+                block,
+                src,
+                "type '{}' could not be memset'ed",
+                .{dest_ptr_ty.fmt(sema.mod)},
+            );
+            errdefer msg.destroy(sema.gpa);
+            try sema.errNote(
+                block,
+                src,
+                msg,
+                "dest must be a mutable slice or a mutable pointer to an array",
```

Comment:
Instead of this error note there should be `"cannot assign to constant"` error which seems to be missing from both `@memcpy`/`@memset`.

```zig
    if (ptr_ty.isConstPtr())
        return sema.fail(block, ptr_src, "cannot assign to constant", .{});
```

Reply:
Hm, the originally suggested error message seems more explicit and clearer to me. If I got that error message I would instantly understand what I needed to do, convert the dest argument from an array to a slice. 

If it just says that it can't to assign to a constant, I would first go "but my array isn't const?" only to then realize that it's getting passed by value and is therefore const. It's just more mental hoops to go through.